### PR TITLE
Tests 2022-Jun

### PR DESCRIPTION
**Pull request by kronenpj**
Set of working tests giving the Python code approximately 70% coverage.

Note: The fix for issue #207 - PR# 208 should be applied for chat text tests to consistently pass.

### CODE REVIEW

**Comment by DJ2LS** [comment]
Wow, Really big improvements 🙂

**Comment by DJ2LS** [comment]
As This is a really big PR I need some more time to review it. 
I reviewed about 75% yet, but I need to have a look at the big changes from my Laptop and with some more time. But everything looks promising 🙂💪🏻

**Comment by kronenpj** [comment]
I'm not fond of the size either. At least it's contained to the tests :).

**Review by DJ2LS — APPROVED** [review]
Really big PR and a lot of work! 

I think we should add a short description to the test files and which additional files belong to it , what exactly we are testing and what's the requirement for passing it. This would possibly make reviewing a little bit easier. But we can add this in a future change after merging this PR and testing the tests / checking how they perform. I think you could also add your name and callsign to the tests files, as its mainly your work and this should be mentioned somewhere!

File: .github/workflows/ctest.yml [hunk 1]
```diff
@@ -18,7 +18,8 @@ jobs:
         run: |
           sudo apt-get update
           sudo apt-get install octave octave-common octave-signal sox python3 python3-pip portaudio19-dev python3-pyaudio
-          pip3 install psutil crcengine ujson pyserial numpy structlog miniaudio sounddevice pytest
+          pip3 install psutil crcengine ujson pyserial numpy structlog miniaudio sounddevice
```
[the remark this answers is not in the document]
Comment (kronenpj) [reply .github/workflows/ctest.yml:21]: It's not used anywhere else. Removing.

File: test/util_tnc_ISS.py [hunk 2]
```diff
@@ -6,33 +6,44 @@
...
+            sock.SOCKET_QUEUE.queue
+        ) and f'"{MESSAGE.lower()}":"sending"' not in str(sock.SOCKET_QUEUE.queue):
             print(f"{MESSAGE} was not sent.")
-            os._exit(1)
+            log.info("iss_arq_cleanup", socket_queue=sock.SOCKET_QUEUE.queue)
+            # sys.exit does not terminate threads, and os_exit doesn't allow coverage collection.
+            signal.raise_signal(signal.SIGKILL)
 
-        os._exit(0)
+        signal.raise_signal(signal.SIGTERM)
```
[the remark this answers is not in the document]
Comment (kronenpj) [reply test/util_tnc_ISS.py:46]: I landed upon this while trying to find a way to terminate the thread while still capturing coverage information. I don't know if Python on Windows supports this as it's originally a POSIX thing which may not translate well to Windows.

File: tnc/data_handler.py [hunk 3]
```diff
@@ -2267,11 +2270,11 @@ def reset_statistics(self) -> None:
         """
         # reset ARQ statistics
         static.ARQ_BYTES_PER_MINUTE_BURST = 0
-        static.ARQ_BYTES_PER_MINUTE = 0
-        static.ARQ_BITS_PER_SECOND_BURST = 0
-        static.ARQ_BITS_PER_SECOND = 0
-        static.ARQ_TRANSMISSION_PERCENT = 0
-        static.TOTAL_BYTES = 0
+        static.ARQ_BYTES_PER_MINUTE       = 0
```
[the remark this answers is not in the document]
Comment (kronenpj) [reply tnc/data_handler.py:2280]: That was unintentional :). I missed that part of the merge. Fixed.

**Comment by kronenpj** [comment]
Unfortunately some of the tests take a while when they fail. When they succeed they're usually quick, but failures need to be allow the receive side enough time to process the data.

**Pull request merged**